velox/exec/Driver.cpp

```diff
@@ -1090,6 +981,71 @@ folly::dynamic Driver::toJson() const {
   return obj;
 }
 
+template <typename Func>
+auto Driver::withDeltaCpuWallTimer(
+    Operator* op,
+    TimingMemberPtr timingMember,
+    Func&& timedFunction) -> decltype(timedFunction()) {
+  // If 'trackOperatorCpuUsage_' is true, create and initialize the timer object
+  // to track cpu and wall time of the timedFunction.
+  if (!trackOperatorCpuUsage_) {
+    return timedFunction();
+  }
+
+  // The delta CpuWallTiming object would be recorded to the corresponding
+  // timingMember upon destruction of the timer when withDeltaCpuWallTimer
+  // ends. The timer is created on the stack to avoid heap allocation
+  auto f = [op, timingMember, this](const CpuWallTiming& elapsedTime) {
+    auto elapsedSelfTime = processLazyTiming(*op, elapsedTime);
+    op->stats().withWLock([&](auto& lockedStats) {
+      (lockedStats.*timingMember).add(elapsedSelfTime);
+    });
+  };
+  DeltaCpuWallTimer<decltype(f)> timer(std::move(f));
+
+  return timedFunction();
+}
+
+void Driver::validateOperatorResult(
+    const RowVectorPtr& result,
+    const Operator& op) {
+  VELOX_CHECK(
+      result->size() > 0,
+      "Operator::getOutput() must return nullptr or a non-empty vector: {}",
+      op.operatorType());
+
+  if (ctx_->queryConfig().validateOutputFromOperators()) {
+    try {
+      result->validate({});
+    } catch (const std::exception& e) {
+      VELOX_FAIL(
+          "Output validation failed for [operator: {}, plan node ID: {}]: {}",
+          op.operatorType(),
+          op.planNodeId(),
+          e.what());
+    }
+  }
+}
+
+StopReason Driver::blockDriver(
+    const std::shared_ptr<Driver>& self,
+    Operator* op,
+    ContinueFuture&& future,
+    std::shared_ptr<BlockingState>& blockingState,
+    CancelGuard& guard) {
+  VELOX_CHECK(
+      future.valid(),
+      "The operator {} is blocked but blocking future is not "
+      "set by isBlocked method.",
+      op->operatorType());
+
+  blockedOperatorId_ = curOperatorId_ + 1;
```

Comment:
It's not always this value, you probably need to pass it in

Reply:
> It's not always this value, you probably need to pass it in

Thanks for the catch! However I don't quite understand why the blocked operator and the blockedOperatorId_ are not the same for the memory arbitration case. Other than this case, the other 3 callsites of blockDriver all use the same blockedOperatorId_ as the operator that's blocked. E.g. if the op is nextOp, then blockedOperatorId_ would be curOperatorId_ + 1; if the op is the current op, the blockedOperatorId_ would be curOperatorId_. The only exception is arbitration:

```
if (FOLLY_UNLIKELY(checkUnderArbitration(&future))) {
          // Blocks the driver if the associated query is under memory
          // arbitration as it is very likely the driver run will trigger memory
          // arbitration when it needs to allocate memory, and the memory
          // arbitration will be blocked by the current running arbitration
          // until it finishes. Instead of blocking the driver thread to wait
          // for the current running arbitration, it is more efficient
          // system-wide to let driver go off thread for the other queries which
          // have free memory capacity to run during the time.
          blockedOperatorId_ = curOperatorId_ + 1;    // blockedOperatorId_ is the next operator's Id
          VELOX_CHECK(future.valid());
          blockingReason_ = BlockingReason::kWaitForArbitration;
          blockingState = std::make_shared<BlockingState>(
              self, std::move(future), op, blockingReason_);   // op is the current operator
          guard.notThrown();
          return StopReason::kBlock;
        }
```
Do you know why this is the case? cc @xiaoxmeng

Reply:
blockedOperatorId_ is mostly for stats reporting. It probably should set to curOperatorId_. @yingsu00 Can you fix that in a separate PR? Thanks!